edb/schema/indexes.py

```diff
@@ -1479,30 +1479,63 @@ def _copy_ext_ai_model_annotations(
 
             self.add(anno_cmd)
 
-        dimensions = model_stype.must_get_json_annotation(
+        model_dimensions = model_stype.must_get_json_annotation(
             schema,
             sn.QualName("ext::ai", "embedding_model_max_output_dimensions"),
             int,
         )
-
         supports_shortening = model_stype.must_get_json_annotation(
             schema,
             sn.QualName("ext::ai", "embedding_model_supports_shortening"),
             bool,
         )
 
+        kwargs = self.scls.get_concrete_kwargs_as_values(schema)
+        specified_dimensions = kwargs["dimensions"]
+
         MAX_DIM = 2000  # pgvector limit
-        if dimensions > MAX_DIM:
-            if not supports_shortening:
+
+        if specified_dimensions is None:
+            if model_dimensions > MAX_DIM:
+                if not supports_shortening:
+                    raise errors.SchemaDefinitionError(
+                        f"{model_stype_vn} returns embeddings with over "
+                        f"{MAX_DIM} dimensions, does not support embedding "
+                        f"shortening, and thus cannot be used with "
+                        f"this index",
+                        span=self.span,
+                    )
+                else:
+                    dimensions = MAX_DIM
+            else:
+                dimensions = model_dimensions
+        else:
+            if specified_dimensions > MAX_DIM:
                 raise errors.SchemaDefinitionError(
-                    f"{model_stype_vn} returns embeddings with over "
-                    f"{MAX_DIM} dimensions, does not support embedding "
-                    f"shortening, and thus cannot be used with "
+                    f"cannot use more than {MAX_DIM} dimensions with "
                     f"this index",
                     span=self.span,
                 )
+            elif specified_dimensions > model_dimensions:
+                raise errors.SchemaDefinitionError(
+                    f"{model_stype_vn} does not support more than "
+                    f"{model_dimensions} dimensions, "
+                    f"got {specified_dimensions}",
+                    span=self.span,
+                )
+            elif (
+                specified_dimensions != model_dimensions
+                and not supports_shortening
+            ):
+                raise errors.SchemaDefinitionError(
+                    f"{model_stype_vn} returns embeddings with over "
```

Comment:
should here just be written: `returns embeddings with ... ` instead of `with over....`?